handshake/src/types/covenant.rs

```diff
@@ -0,0 +1,1007 @@
+
+        for covenant_data in self.0.clone() {
+            total += covenant_data.write_to(writer)?;
+        }
+
+        Ok(total)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+/// A Covenant is a part of a Handshake UTXO that restricts how the UTXO may
+/// be spent over the period of multiple spends. It is able to maintain
+/// data over time.
+pub struct Covenant {
+    /// The CovenantType restricts spending to other CovenantTypes that have
+    /// a defined spending path. State machines can be constructed this way
+    /// that enable the auction system.
+    pub covenant_type: CovenantType,
+    /// The CovenantData represents Covenant specific data that is attached to
+    /// the UTXO. This data must be updated correctly as the UTXO is spent from
+    /// one CovenantType to the next CovenantType.
+    pub covenant_data: CovenantData,
+}
+
+impl Covenant {
+    /// Returns the null Covenant.
+    pub fn null() -> Self {
+        Self {
+            covenant_type: CovenantType::NONE,
+            covenant_data: CovenantData::null(),
+        }
+    }
+}
+
+impl ByteFormat for Covenant {
+    type Error = SerError;
+
+    fn serialized_length(&self) -> usize {
+        let mut size: usize = 1; // covenant_type
+        size += self.covenant_data.serialized_length();
+        size
+    }
+
+    fn read_from<R>(reader: &mut R, _limit: usize) -> SerResult<Self>
+    where
+        R: Read,
+        Self: std::marker::Sized,
+    {
+        let mut buf = [0u8; 1];
+        reader.read_exact(&mut buf)?;
+        let covenant_type = u8::from_le_bytes(buf);
+        // TODO(mark): sane max?
+        let covenant_data = CovenantData::read_from(reader, 1024)?;
+
+        Ok(Self {
+            covenant_type: covenant_type.into(),
+            covenant_data,
+        })
+    }
+
+    fn write_to<W>(&self, writer: &mut W) -> SerResult<usize>
+    where
+        W: Write,
+    {
+        let mut total: usize = 0;
+        let covenant_type: u8 = self.covenant_type.clone() as u8;
+        total += writer.write(&[covenant_type])?;
+        total += &self.covenant_data.write_to(writer)?;
+
+        Ok(total)
+    }
+}
+
+/// Covenant types
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+#[repr(u8)]
+pub enum CovenantType {
+    /// Standard Bitcoin-like Output
+    NONE = 0,
+    /// Reserved Name Claim Output
+    CLAIM = 1,
+    /// Auction Opening Output
+    OPEN = 2,
+    /// Blinded Bid Output
+    BID = 3,
+    /// Reveal Output
+    REVEAL = 4,
+    /// Auction Losers Must Redeem
+    REDEEM = 5,
+    /// Auction Winners Must Register
+    REGISTER = 6,
+    /// Update Name State
+    UPDATE = 7,
+    /// Renew Name
+    RENEW = 8,
+    /// Begin to Transfer Name
+    TRANSFER = 9,
+    /// Finalize Name Transfer
+    FINALIZE = 10,
+    /// Burn Name When Key is Compromised
+    REVOKE = 11,
+    #[doc(hidden)]
+    UNKNOWN12,
```

Comment:
I died a little bit on the inside while doing this but it seemed like a good idea at the time since its consensus critical. I read that doing it with a macro wouldn't be easy so each possible variant of a covenant type (0-255) exists in this enum. It ensures that undefined covenant values are passed through when serialized.

Reply:
What about `pub struct CovenantType(u8)`? might result in less terrible code?

Reply:
Using a newtype like that seems like a far better idea